### PR TITLE
TST: Test key dtype cast after merge

### LINKED ISSUE #17044: Type casting after merge
#### Code Sample, a copy-pastable example if possible

```python
In [2]: df1 = pd.DataFrame({'id':[1.,2.], 'c1':[10,20]})

In [3]: df2 = pd.DataFrame({'id':[2], 'c2':[200]})

In [4]: df = df1.merge(df2, on='id', how='left')

In [5]: df['id'].dtype
Out[5]: dtype('O')
```
#### Problem description

When I merge 2 DataFrames on a column (`id`) which is of type float in one of the DataFrames and of type int in the other DataFrame, the resulting DataFrame column is of type Object. Why? I guess the resulting DataFrame should keep the float type.

#### Expected Output
```python
In [5]: df['id'].dtype
Out[5]: dtype('float64')
```
Indeed, this was the behavior for pandas 0.19.2.

#### Output of ``pd.show_versions()``

<details>
INSTALLED VERSIONS
------------------
commit: None
python: 3.6.1.final.0
python-bits: 64
OS: Linux
OS-release: 2.6.32-431.29.2.el6.x86_64
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: C
LOCALE: None.None

pandas: 0.20.2
pytest: 3.0.7
pip: 9.0.1
setuptools: 27.2.0
Cython: 0.25.2
numpy: 1.13.1
scipy: 0.19.1
xarray: None
IPython: 6.1.0
sphinx: 1.5.6
patsy: 0.4.1
dateutil: 2.6.0
pytz: 2017.2
blosc: None
bottleneck: 1.2.1
tables: 3.3.0
numexpr: 2.6.2
feather: None
matplotlib: 2.0.2
openpyxl: 2.4.7
xlrd: 1.0.0
xlwt: 1.2.0
xlsxwriter: 0.9.6
lxml: 3.7.3
bs4: 4.6.0
html5lib: 0.999
sqlalchemy: 1.1.9
pymysql: None
psycopg2: None
jinja2: 2.9.6
s3fs: None
pandas_gbq: None
pandas_datareader: None
</details>

### PR DESCRIPTION
**Pull request by albertvillanova**
- [X] closes #17044
- [x] tests added / passed
- [x] passes `black pandas`
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

**Comment by jbrockmendel** [comment]
Travis fail is likely unrelated, but there are some linting issues you need to address

File: pandas/tests/reshape/merge/test_merge.py [hunk 1]
```diff
@@ -1437,6 +1437,13 @@ def test_merge_on_ints_floats(self, int_vals, float_vals, exp_vals):
         result = B.merge(A, left_on="Y", right_on="X")
         assert_frame_equal(result, expected[["Y", "X"]])
 
+    def test_merge_key_dtype_cast(self):
+        # GH 17044
+        df1 = pd.DataFrame({'key': [1., 2.], 'c1': [10, 20]})
+        df2 = pd.DataFrame({'key': [2], 'c2': [200]})
+        df = df1.merge(df2, on='key', how='left')
+        assert df['key'].dtype == 'float64'
```
Comment (mroeschke) [inline pandas/tests/reshape/merge/test_merge.py]: If you could `tm.assert_frame_equal` the resulting merged dataframe, that would be ideal (which checks the dtypes too.

```
result = df1.merge(df2, on='key', how='left')
expected = ....
tm.assert_frame_equal(result, expected)

```

File: pandas/tests/reshape/merge/test_merge.py [hunk 2]
```diff
@@ -1437,6 +1437,15 @@ def test_merge_on_ints_floats(self, int_vals, float_vals, exp_vals):
         result = B.merge(A, left_on="Y", right_on="X")
         assert_frame_equal(result, expected[["Y", "X"]])
 
+    def test_merge_key_dtype_cast(self):
+        # GH 17044
+        df1 = DataFrame({'key': [1., 2.], 'c1': [10, 20]})
+        df2 = DataFrame({'key': [2], 'c2': [200]})
+        result = df1.merge(df2, on='key', how='left')
+        expected = DataFrame({'key': [1.0, 2.0], 'c1': [10, 20], 'c2': [np.nan, 200.0]})
+        # df['key'].dtype == 'float64'
```
Comment (mroeschke) [inline pandas/tests/reshape/merge/test_merge.py]: Could you remove this comment?

**Comment by mroeschke** [comment]
Could you merge master? Master was broken recently, but it should be stable now.

File: pandas/tests/reshape/merge/test_merge.py [hunk 3]
```diff
@@ -1437,6 +1437,14 @@ def test_merge_on_ints_floats(self, int_vals, float_vals, exp_vals):
         result = B.merge(A, left_on="Y", right_on="X")
         assert_frame_equal(result, expected[["Y", "X"]])
 
+    def test_merge_key_dtype_cast(self):
+        # GH 17044
+        df1 = DataFrame({'key': [1., 2.], 'c1': [10, 20]})
```
Comment (TomAugspurger) [inline pandas/tests/reshape/merge/test_merge.py]: Need to specify `columns` for the Py35 build, else the dict is sorted.

**Comment by mroeschke** [comment]
Just need to run `black` otherwise LGTM.

**Review by mroeschke — APPROVED** [review]

**Review by WillAyd — APPROVED** [review]

**Comment by WillAyd** [comment]
Thanks @albertvillanova

**Pull request merged**